pageserver/src/tenant/timeline.rs

```diff
@@ -5628,30 +5619,36 @@ impl Timeline {
 
         let distance_based_decision = distance.0 >= min_distance;
 
-        let mut time_based_decision = false;
         let mut last_check_instant = self.last_image_layer_creation_check_instant.lock().unwrap();
+        let mut check_required_after = self.get_checkpoint_timeout();
+
         if let CurrentLogicalSize::Exact(logical_size) = self.current_logical_size.current_size() {
-            let check_required_after =
+            check_required_after =
                 if Some(Into::<u64>::into(&logical_size)) >= large_timeline_threshold {
                     self.get_checkpoint_timeout()
                 } else {
                     Duration::from_secs(3600 * 48)
                 };
-
-            time_based_decision = match *last_check_instant {
-                Some(last_check) => {
-                    let elapsed = last_check.elapsed();
-                    elapsed >= check_required_after
-                }
-                None => true,
-            };
         }
+        // HADRON: we removed the check on self.current_logical_size non shard-0 because only shard 0 computes exact size.
```

Comment:
Hm. I still see the check above. The issue is legitimate (and slightly embarassing 😓).

The context for this change was that neon deployments have very large numbers of timelines (sometimes more than 10k on one PS) and we were burning a lot of CPU time on the image layer check.

I propose we keep the check, but only for unsharded tenants. If the tenant is sharded, there should be at least one timeline with logical size greater than 16GiB (that's the initial split threshold). It's not a perfect check, but should keep things stable since most of the fleet is unsharded.

To clarify, im proposing smth like:
```
check_required_after = if logical_size > large_timeline_threhsold || self_is_sharded {
  self.get_checkpoint_timeout()
} else {
  Duration::from_secs(3600 * 48)
};
```

Reply:
Done in [e6e7c3f](https://github.com/neondatabase/neon/pull/12552/commits/e6e7c3f33dd068ca90d70346fd3bfdb796dab3ef)